File: examples_C++/maxwell1D.cpp
/**
 * This example uses MOLE to solve 1D Maxwell's equations in free space.
 */

#include "mole.h"
#include <iostream>

int main() {    
    int nt = 50;            // Number of time steps
    int m = 100;            // Number of cells
    int k = 2;              // Operators' order of accuracy
    Real a = 0;             // Left boundary
    Real b = 1;             // Right boundary
    Real dx = (b - a) / m;  // Step size
    Real dt = 0.5 * dx;     // CFL-based timestep

    // Get mimetic operators
    Divergence D(k, m, dx);
    D *= dt;
    Gradient G(k, m, dx);
    G *= dt;

    // Grids
    vec centers(m + 2);
    centers(0) = a;
    centers(1) = a + dx / 2.0;
    int i;
    for (i = 2; i <= m; i++)
        centers(i) = centers(i - 1) + dx;
    centers(i) = b;

    // Initialize fields
    vec B = zeros<vec>(m + 1);
    vec E = exp(-100.0 * square(centers - 0.5));  // Gaussian pulse

    // Leapfrog start: half-step back for B
    B -= 0.5 * (G * E);

    // Time stepping
    for (int n = 0; n < nt; ++n) {
        // Update B (edges)
        B -= G * E;

        // Store boundary values
        Real El = E(1);
        Real Er = E(m);

        // Update E (centers)
        E -= D * B;

        // Absorbing boundaries
        E(0)     = El;
        E(m + 1) = Er;
    }

    vec edges = linspace(a, b, m + 1);

    // Dump fields to stdout
    for (size_t i = 0; i < centers.n_elem; ++i)
        cout << centers(i) << " " << E(i) << endl;

    cout << endl;

    for (size_t i = 0; i < edges.n_elem; ++i)
        cout << edges(i) << " " << B(i) << endl;

    return 0;
}

File: examples_MATLAB/maxwell1D.m
% Solution to 1D Maxwell's equations using mimetic operators and leapfrog
clear; clc; close all;

addpath("../mole_MATLAB");

nt = 150;         % number of time steps
m = 100;          % number of cells
east = 1;         % domain's limits
west = 0;
dx = (east - west)/m;
dt = 0.5*dx;      % time step based on CFL

% Grids
xE = [west west+dx/2 : dx : east-dx/2 east];
xB = west : dx : east;

% Mimetic operators
k = 2; % Order of accuracy
D = dt*div(k, m, dx);
G = dt*grad(k, m, dx);

% Initialize magnetic field
B = zeros(m+1, 1);

% Initial condition: Gaussian pulse in E
E = exp(-100*(xE-0.5).^2).';

% Leapfrog: stagger B by half step
B = B - 0.5*G*E;  % half-step backward for leapfrog start

figure;

% Time-stepping
for n = 1:nt
    % Update B (edges)
    B = B - G*E;

    % Store bdry values
    El = E(2);
    Er = E(end-1);
    
    % Update E (centers)
    E = E - D*B;

    % Absorbing boundary condition
    E(1)   = El;    % left boundary
    E(end) = Er;    % right boundary
    
    % Plot E field (subplot 1)
    subplot(2, 1, 1);
    plot(xE, E, 'b-', 'LineWidth', 2);
    title('Electric Field');
    xlabel('x');
    ylabel('E(x,t)');
    xlim([0 1]);
    ylim([-1 1]);
    grid on;

    % Plot B field (subplot 2)
    subplot(2, 1, 2);
    plot(xB, B, 'r-', 'LineWidth', 2);
    title('Magnetic Field');
    xlabel('x');
    ylabel('B(x,t)');
    xlim([0 1]);
    ylim([-1 1]);
    grid on;

    pause(0.01)
end
